provider/newrelic: Add a New Relic provider

paultyng opened the pull request:
This is still a work in progress, but I wanted to get some feedback / thoughts on if this is desirable and ways to improve it as its my first larger contribution to terraform.

The provider allows you to manage your New Relic alert policies, notification channels, alert conditions, etc.

Coupled with the Pager Duty provider, you can even link up alerts to pager duty solely in terraform.

(Ignore the commit messages, I'll clean it up all up once its closer to complete)

## Update: Running Acceptance Tests

To run acceptance tests, you'll need to setup a new New Relic account, and generate an **Admin** API key.  You will also need the license key, only for acceptance tests though, its not used in the actual provider.  

You can then pass the API key via the `NEWRELIC_API_KEY` environment var and the license key via `NEWRELIC_LICENSE_KEY`

```bash
NEWRELIC_API_KEY=1234567 NEWRELIC_LICENSE_KEY=8901234 make testacc TEST=./builtin/providers/newrelic
```

paultyng commented:
The main issue was there was new good Go API client for New Relic alerts.  So I've also been writing one of those as well.  Their API is unfortunately pretty limited and not fully featured, so I've had to work around some of those in this implementation.  They also only provide a Swagger 1.1 spec, and I couldn't get a great autogenerated client from it in any tools I tried, so I just went with hand coding for now.

stack72 commented:
Hey @paultyng 

This is an exciting provider - ability to just bootstrap systems like this is awesome. In fairness, this is going to be a bit of a beast to review. It's a single commit with 225 files so far :)

Can you break this up a bit? Something like a commit per resource and a commit for the SDK addition would be really awesome!

Thoughts?

Paul

paultyng commented:
Yeah no problem, I was planning on rewriting the commits once I got this closer to an MVP, so I'll do it in the way you suggest.

stack72 commented:
Awesome! That will help a lot :)

P.

paultyng commented:
@stack72 Updated with better commit messages and I think a good set of MVP resources.  There are still some annoyances where you need magic ID's from new relic to make it work.  I can add data sources to help in looking those up (like applications you have configured, etc).  As far as I can tell though you can't create them via API so they can't be resources.  Not sure exactly how much I should add to the PR as it will probably just make reviewing a pain.

tiny-dancer commented:
very exciting, awesome work @paultyng! on a slightly similar note, did you by chance take a look at the feasibility of integrating with the synthetics api?
 [Synthetics API Docs](https://docs.newrelic.com/docs/apis/synthetics-rest-api/monitor-examples/manage-synthetics-monitors-via-rest-api#sample-script )

paultyng commented:
I've added some `ValidationFunc`s in to help with attributes that are lists of keys.

paultyng commented:
@tiny-dancer I don't currently use the Synthetics API so wasn't in my initial plan.  I want to try to keep this PR small but useful, but once the provider is in, adding additional functionality should be a bit more streamlined and can look in to it then.

tiny-dancer commented:
@paultyng sound good. and yah wasn't thinking for this PR, just curious if it was in your plans. was about to implement a temp solution with jenkinsfile to bring source control to the pingers/scripted tests until i ran across this pr.  if we can build off the start here, makes muuch more sense in terraform.  on the surface would seem to build nicely for directly relating the policies defined here to the synthetic tests. 👍

paultyng commented:
@stack72 I'm not sure what to do in the acceptance tests about the data source testing, as it requires manually creating data in New Relic and testing to see if its present. ([see here](https://github.com/hashicorp/terraform/pull/10317/files#diff-fdb9faa183b21057d06ec7d8814d5440R44)).  

I guess its possible I could fake creating an application by sending a metric ping to new relic, not sure if that works or not, I guess I can test it out.

paultyng commented:
[Seems like it will work](https://docs.newrelic.com/docs/agents/go-agent/get-started/get-new-relic-go), but it will require their go agent and the reporting api key for the acceptance tests, but seems a lot nicer than having to configure it in the UI for tests to pass.

paultyng commented:
Ok, added app creation via the agent sdk just for testing.  Can maybe eventually work that in to a resource instead of a data source but it requires a separate API key and integration, so a little annoying.  I updated my acceptance test instructions above.

paultyng commented:
@stack72 need anything from me on this? Anyone available for a code review?

stack72 commented:
Hi @paultyng 

is this no longer considered a work in progress now? 

Paul

stack72 commented on builtin/providers/newrelic/data_source_newrelic_application_test.go:
```diff
@@ -0,0 +1,48 @@
+// This name must match the constant defined in provider_test.go
+const testAccNewRelicApplicationConfig = `
+data "newrelic_application" "app" {
+	name = "terraform-test-application"
```
We should randomise this application name - this would stop any duplicate resource names

An example can be found [here](https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/resource_aws_iam_instance_profile_test.go#L160)

paultyng commented:
Correct, I think this has enough for v1.

stack72 commented on builtin/providers/newrelic/resource_newrelic_alert_channel.go:
```diff
@@ -0,0 +1,167 @@
+
+	d.Set("name", channel.Name)
+	d.Set("type", channel.Type)
+	d.Set("configuration", channel.Configuration)
```
As this is a complex structure - I think we should check for an error setting the value, e.g.

```
if err := d.Set("configuration", channel.Configuration); err != nil {
			return fmt.Errorf("[DEBUG] Error setting Alert Channel Configuration: %#v", err)
		}
```

stack72 commented on builtin/providers/newrelic/resource_newrelic_alert_channel.go:
```diff
@@ -0,0 +1,167 @@
+	return nil
+}
+
+func resourceNewRelicAlertChannelImport(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
```
As we don't have any custom logic in this import, we should just add it to the schema as follows:

```
Importer: &schema.ResourceImporter{
			State: schema.ImportStatePassthrough,
		},
```

This calls the Read func by default :)

stack72 commented on builtin/providers/newrelic/resource_newrelic_alert_channel_test.go:
```diff
@@ -0,0 +1,125 @@
+
+const testAccCheckNewRelicAlertChannelConfig = `
+resource "newrelic_alert_channel" "foo" {
+  name = "foo"
```
Should be randomized to avoid test collision

stack72 commented on builtin/providers/newrelic/resource_newrelic_alert_channel_test.go:
```diff
@@ -0,0 +1,125 @@
+
+const testAccCheckNewRelicAlertChannelConfigUpdated = `
+resource "newrelic_alert_channel" "foo" {
+  name = "bar"
```
Should be randomized to avoid test collision

stack72 commented on builtin/providers/newrelic/resource_newrelic_alert_condition.go:
```diff
@@ -0,0 +1,338 @@
+		Update: resourceNewRelicAlertConditionUpdate,
+		Delete: resourceNewRelicAlertConditionDelete,
+		Importer: &schema.ResourceImporter{
+			State: resourceNewRelicAlertConditionImport,
```
This can just be:

```
State: schema.ImportStatePassthrough,
```

stack72 commented on builtin/providers/newrelic/resource_newrelic_alert_condition.go:
```diff
@@ -0,0 +1,338 @@
+		terms = append(terms, dst)
+	}
+
+	d.Set("term", terms)
```
We should maybe log an error here if we cannot set terms to state (as described previously)

stack72 commented on builtin/providers/newrelic/resource_newrelic_alert_condition.go:
```diff
@@ -0,0 +1,338 @@
+	policyID := ids[0]
+	id := ids[1]
+
+	log.Printf("[INFO] Deleting New Relic alert condition %v", id)
```
is %v correct for this? Wouldn't %s be better as it's a string?

paultyng replied:
Its an int in this case

paultyng replied:
I can switch to `%d`?

stack72 commented on builtin/providers/newrelic/resource_newrelic_alert_condition_test.go:
```diff
@@ -0,0 +1,183 @@
+
+const testAccCheckNewRelicAlertConditionConfig = `
+resource "newrelic_alert_policy" "foo" {
+  name = "foo"
```
Randomize the test name :)

stack72 commented on builtin/providers/newrelic/resource_newrelic_alert_condition_test.go:
```diff
@@ -0,0 +1,183 @@
+resource "newrelic_alert_condition" "foo" {
+  policy_id = "${newrelic_alert_policy.foo.id}"
+
+  name        = "foo"
```
randomize the test name :)

stack72 commented on builtin/providers/newrelic/resource_newrelic_alert_condition_test.go:
```diff
@@ -0,0 +1,183 @@
+
+const testAccCheckNewRelicAlertConditionConfigUpdated = `
+resource "newrelic_alert_policy" "foo" {
+  name = "foo"
```
randomize

stack72 commented on builtin/providers/newrelic/resource_newrelic_alert_policy.go:
```diff
@@ -0,0 +1,122 @@
+		// Update: Not currently supported in API
+		Delete: resourceNewRelicAlertPolicyDelete,
+		Importer: &schema.ResourceImporter{
+			State: resourceNewRelicAlertPolicyImport,
```
Can just be a PassThrough

paultyng commented:
Most  (if not all) of the names do not have unique constraints in new relic.  Should I still randomize?

paultyng commented:
@stack72 changes made, I went ahead and added the randomization to the names anyway.

paultyng commented:
@stack72 Let me know if there are some additional changes I should make.

paultyng commented:
I imagine you are all very busy with 0.8, let me know if there is anything I can do before you take another pass at this to help expedite it.

stack72 commented:
hi @paultyng 

Coming back to look at this again :) Give me like 30 mins and I will know the state of what is going on with it

Paul

stack72 commented:
Hi @paultyng 

This is looking wonderful now :) The tests all pass and I added the documentation link to the nav bar

```
% make testacc TEST=./builtin/providers/newrelic
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/12/15 19:07:45 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/newrelic -v  -timeout 120m
=== RUN   TestAccNewRelicApplication_Basic
--- PASS: TestAccNewRelicApplication_Basic (4.79s)
=== RUN   TestParseIDs_Basic
--- PASS: TestParseIDs_Basic (0.00s)
=== RUN   TestSerializeIDs_Basic
--- PASS: TestSerializeIDs_Basic (0.00s)
=== RUN   TestAccNewRelicAlertChannel_import
--- PASS: TestAccNewRelicAlertChannel_import (4.48s)
=== RUN   TestAccNewRelicAlertCondition_import
--- PASS: TestAccNewRelicAlertCondition_import (4.72s)
=== RUN   TestAccNewRelicAlertPolicy_import
--- PASS: TestAccNewRelicAlertPolicy_import (3.84s)
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProviderImpl
--- PASS: TestProviderImpl (0.00s)
=== RUN   TestAccNewRelicAlertChannel_Basic
--- PASS: TestAccNewRelicAlertChannel_Basic (9.13s)
=== RUN   TestAccNewRelicAlertCondition_Basic
--- PASS: TestAccNewRelicAlertCondition_Basic (7.91s)
=== RUN   TestAccNewRelicAlertPolicyChannel_Basic
--- PASS: TestAccNewRelicAlertPolicyChannel_Basic (9.83s)
=== RUN   TestAccNewRelicAlertPolicy_Basic
--- PASS: TestAccNewRelicAlertPolicy_Basic (5.67s)
=== RUN   TestValidationIntInInSlice
--- PASS: TestValidationIntInInSlice (0.00s)
=== RUN   TestValidationFloat64Gte
--- PASS: TestValidationFloat64Gte (0.00s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/newrelic	50.390s
```

Going to manually merge this as I am rebasing into 3 logical commits. Thanks for all the work you have put in here - it's great!

Paul
